Type error when using test.each with a class starting with types/bun 1.2.23

### What version of Bun is running?

1.2.23+cf1367137

### What platform is your computer?

Darwin 24.6.0 arm64 arm

### What steps can reproduce the bug?

Pretty sure we can reproduce this with something simpler.

Using `"@types/bun": "1.2.23"`. With `1.2.22` it works.

You can simply `bun init -y` and paste this code in a `repro.test.ts file`

```ts
import { describe, expect, test } from "bun:test";

class A {}

describe("bun", () => {
	const values = [
		["hello", "world"],
		["hello", new A()], 
	] as const;
	describe("format", () => {
		test.each(values)("value", (_builder, _expected) => {
			expect(true).toBe(true);
		});
	});
});
```



### What is the expected behavior?

same file with 1.2.22 gives no type error.

### What do you see instead?

```
repro.test.ts:11:30 - error TS2345: Argument of type '(_builder: "hello", _expected: A | "world") => void' is not assignable to parameter of type '(...args: ["hello", "world", (err?: unknown) => void] | ["hello", A, (err?: unknown) => void]) => void | Promise<unknown>'.
  Types of parameters '_builder' and 'args' are incompatible.
    Type '["hello", "world", (err?: unknown) => void] | ["hello", A, (err?: unknown) => void]' is not assignable to type '[_builder: "hello", _expected: A | "world"]'.
      Type '["hello", "world", (err?: unknown) => void]' is not assignable to type '[_builder: "hello", _expected: A | "world"]'.
        Source has 3 element(s) but target allows only 2.

11   test.each(values)("value", (_builder, _expected) => {
                                ~~~~~~~~~~~~~~~~~~~~~~~~~~


Found 1 error in repro.test.ts:11
```

### Additional information

Having a class in the values and more than one value seems to cause the issue. 

Quite specific 🤔  so not sure if there's a simpler reproduction or not

--- a/packages/bun-types/test.d.ts
+++ b/packages/bun-types/test.d.ts
@@ -390,11 +390,20 @@ declare module "bun:test" {
      */
     repeats?: number;
   }
-  type IsTuple<T> = T extends readonly unknown[]
-    ? number extends T["length"]
-      ? false // It's an array with unknown length, not a tuple
-      : true // It's an array with a fixed length (a tuple)
-    : false; // Not an array at all
+
+  namespace __internal {
+    type IsTuple<T> = T extends readonly unknown[]
+      ? number extends T["length"]
+        ? false // It's an array with unknown length, not a tuple
+        : true // It's an array with a fixed length (a tuple)
+      : false; // Not an array at all
+
+    /**
+     * Accepts `[1, 2, 3] | ["a", "b", "c"]` and returns `[1 | "a", 2 | "b", 3 | "c"]`
+     */
+    type Flatten<T, Copy extends T = T> = { [Key in keyof T]: Copy[Key] };
+  }
+
   /**
    * Runs a test.
    *
@@ -418,10 +427,16 @@ declare module "bun:test" {
    *
    * @category Testing
    */
-  export interface Test<T extends Readonly<any[]>> {
+  export interface Test<T extends ReadonlyArray<unknown>> {
     (
       label: string,
-      fn: (...args: IsTuple<T> extends true ? [...T, (err?: unknown) => void] : T) => void | Promise<unknown>,
+
+      fn: (
+        ...args: __internal.IsTuple<T> extends true
+          ? [...table: __internal.Flatten<T>, done: (err?: unknown) => void]
+          : T
+      ) => void | Promise<unknown>,
+
       /**
        * - If a `number`, sets the timeout for the test in milliseconds.
        * - If an `object`, sets the options for the test.
@@ -513,8 +528,8 @@ declare module "bun:test" {
      *
      * @param table Array of Arrays with the arguments that are passed into the test fn for each row.
      */
-    each<T extends Readonly<[any, ...any[]]>>(table: readonly T[]): Test<[...T]>;
-    each<T extends any[]>(table: readonly T[]): Test<[...T]>;
+    each<T extends Readonly<[unknown, ...unknown[]]>>(table: readonly T[]): Test<[...T]>;
+    each<T extends unknown[]>(table: readonly T[]): Test<[...T]>;
     each<T>(table: T[]): Test<[T]>;
   }
   /**

--- a/test/integration/bun-types/fixture/23347.test.ts
+++ b/test/integration/bun-types/fixture/23347.test.ts
@@ -1,0 +1,31 @@
+import { expect, test } from "bun:test";
+
+const eachWithDifferingTypes = test.each([
+  ["hello", "world"],
+  ["hello", 2],
+]);
+
+eachWithDifferingTypes("each with differing types", (a, b) => {
+  expect(typeof a).toBe("string");
+  expect(["world", 2]).toContain(b);
+});
+
+const eachWithConst = test.each([
+  ["hello", "world"],
+  ["hello", "alistair"],
+] as const);
+
+eachWithConst("each with `as const`", (a, b) => {
+  expect(a).toBe("hello");
+  expect(["world", "alistair"]).toContain(b);
+});
+
+const eachWithConstAndDifferingTypes = test.each([
+  ["hello", "world"],
+  ["hello", 2],
+] as const);
+
+eachWithConstAndDifferingTypes("each with `as const` and differing types", (a, b) => {
+  expect(a).toBe("hello");
+  expect(["world", 2]).toContain(b);
+});

--- a/test/integration/bun-types/fixture/test.ts
+++ b/test/integration/bun-types/fixture/test.ts
@@ -122,6 +122,9 @@ const data = [
   ["a", true, 5],
   ["b", false, "asdf"],
 ];
+
+const a = test.each(data);
+
 test.each(data)("test.each", (a, b, c) => {
   expectType<string | number | boolean | ((err?: unknown) => void)>(a);
   expectType<string | number | boolean | ((err?: unknown) => void)>(b);

--- a/test/integration/bun-types/fixture/test.ts
+++ b/test/integration/bun-types/fixture/test.ts
@@ -123,8 +123,6 @@ const data = [
   ["b", false, "asdf"],
 ];
 
-const a = test.each(data);
-
 test.each(data)("test.each", (a, b, c) => {
   expectType<string | number | boolean | ((err?: unknown) => void)>(a);
   expectType<string | number | boolean | ((err?: unknown) => void)>(b);

